### ISSUE TITLE
CampaignStatus 'In Progress' not supported

### ISSUE THREAD

**Issue by tmacht** [Jericho/StrongGrid#194]
See stack trace:
`TypeNewtonsoft.Json.JsonSerializationExceptionMessageError converting value "In Progress" to type 'StrongGrid.Model.CampaignStatus'. Path 'status', line 1, position 17658.StackTrace   at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType, JsonSerializer jsonSerializer)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject[T]()    at StrongGrid.Utilities.Extensions.d__12`1.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()    at StrongGrid.Utilities.Extensions.d__7`1.MoveNext()TypeSystem.ArgumentExceptionMessageRequested value 'In Progress' was not found.StackTrace   at System.Enum.EnumResult.SetFailure(ParseFailureKind failure, String failureMessageID, Object failureMessageFormatArgument)    at System.Enum.TryParseEnum(Type enumType, String value, Boolean ignoreCase, EnumResult& parseResult)    at System.Enum.Parse(Type enumType, String value, Boolean ignoreCase)    at Newtonsoft.Json.Utilities.EnumUtils.ParseEnumName(String enumText, Boolean isNullable, Type t)    at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer) | Type | Newtonsoft.Json.JsonSerializationException | Message | Error converting value "In Progress" to type 'StrongGrid.Model.CampaignStatus'. Path 'status', line 1, position 17658. | StackTrace | at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType, JsonSerializer jsonSerializer)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject[T]()    at StrongGrid.Utilities.Extensions.d__12`1.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()    at StrongGrid.Utilities.Extensions.d__7`1.MoveNext() | Type | System.ArgumentException | Message | Requested value 'In Progress' was not found. | StackTrace | at System.Enum.EnumResult.SetFailure(ParseFailureKind failure, String failureMessageID, Object failureMessageFormatArgument)    at System.Enum.TryParseEnum(Type enumType, String value, Boolean ignoreCase, EnumResult& parseResult)    at System.Enum.Parse(Type enumType, String value, Boolean ignoreCase)    at Newtonsoft.Json.Utilities.EnumUtils.ParseEnumName(String enumText, Boolean isNullable, Type t)    at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Type | Newtonsoft.Json.JsonSerializationException
Message | Error converting value "In Progress" to type 'StrongGrid.Model.CampaignStatus'. Path 'status', line 1, position 17658.
StackTrace | at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateObject(Object newObject, JsonReader reader, JsonObjectContract contract, JsonProperty member, String id)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType, JsonSerializer jsonSerializer)    at Newtonsoft.Json.Linq.JToken.ToObject(Type objectType)    at Newtonsoft.Json.Linq.JToken.ToObject[T]()    at StrongGrid.Utilities.Extensions.d__12`1.MoveNext()  --- End of stack trace from previous location where exception was thrown ---    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)    at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()    at StrongGrid.Utilities.Extensions.d__7`1.MoveNext()
Type | System.ArgumentException
Message | Requested value 'In Progress' was not found.
StackTrace | at System.Enum.EnumResult.SetFailure(ParseFailureKind failure, String failureMessageID, Object failureMessageFormatArgument)    at System.Enum.TryParseEnum(Type enumType, String value, Boolean ignoreCase, EnumResult& parseResult)    at System.Enum.Parse(Type enumType, String value, Boolean ignoreCase)    at Newtonsoft.Json.Utilities.EnumUtils.ParseEnumName(String enumText, Boolean isNullable, Type t)    at Newtonsoft.Json.Converters.StringEnumConverter.ReadJson(JsonReader reader, Type objectType, Object existingValue, JsonSerializer serializer)`

**Comment by Jericho** [comment]
This is interesting. I'm guessing you were retrieving a campaign when this error occurred? Any chance you can paste the JSON that SendGrid returned?

What's also interesting is that the status mentioned in the error message is `In Progress` with certain letters capitalized. My experience is that SendGrid normally uses all lowercase like: `draft`, `scheduled`, `sent`, etc.

**Comment by tmacht** [comment]
Yes, this happened when retrieving a campaign. Unfortunately I do not have
the JSON from SendGrid, since we do not log that. I suspect that this
happens when a campaign is cancelled. I will try next week to reproduce the
situation and come back to you.

**Comment by tmacht** [comment]
Hi,

Sorry, it took me longer to come back to this. I can now confirm that
SendGrid returns a campaign with status "In Progress" if the campaign is
being sent right now. This is not so easy to reproduce, because you have to
retrieve the campaign at the right moment and sending is very quick. I
managed to reproduce this by retrieving the campaign in an infinite loop
which increases the chance to retrieve the campaign at the right moment.

I found that the CampaignStatus 'Canceled' is also not supported by
StrongGrid. This is even harder to reproduce, because you can only cancel a
campaign while it is being sent. I did not manage to reproduce this on a
test environment, because I always missed the right moment. So I am not
exactly sure if SendGrid returns 'Canceled', or 'canceled' or 'cancelled'
or something else ... Maybe you have an idea how to test this?

Regards,
Thomas Macht

**Comment by Jericho** [comment]
I asked SendGrid if they could provide a list of possible statuses but I haven't heard back yet.

In the mean time, I have been conducting a few tests and confirmed that conversion from string to enum is case insensitive which means, for example, that `scheduled`, `Scheduled`, `SCHEDULED` will all convert into `CampaignStatus.Scheduled`. For some reason I was under the impression that statuses returned by SendGrid were all lower case so I was surprised when you reported "In Progress".

As far as ideas to help you capture the data returned by SendGrid and determine the exact spelling of the statuses, yes I have a suggestion. You currently probably have a line similar to this: `var client = new Client(apiKey);`. Replace it with: `var client = new Client(apiKey, new System.Net.WebProxy("http://localhost:8888"));` and run [fiddler](https://www.telerik.com/download/fiddler) right before you start your testing code with the infinite loop.

When fiddler is started, click on the "File" menu and make sure the "Capture Traffic" option is unchecked. This will ensure fiddler doesn't capture from your browsers, vpn, email client or any other HTTP tool which could be overwhelming.

In the fiddler UI, look for a button called "Inspectors":
![image](https://user-images.githubusercontent.com/112710/31177473-245fa5c4-a8e4-11e7-81a4-49e5b3c944b7.png)

Clicking on this button will reveal two panels: the first panel shows the requests you send to SendGrid and the second panel shows responses. Please note there are a dozen or so buttons above each panels such as text view, syntax view, web forms, etc to control how the data is presented to you. I personally prefer the "raw" option.

When your "infinite loop" errors out, you'll be able to go back to fiddler, look at the most recent request/response and copy the data in the response. This will allow us to get the exact spelling of the campaign status.

**Comment by Jericho** [comment]
Ok, someone did respond to my question and evidently the various campaign statuses are: Sent, Draft, In Progress, Canceled and Scheduled. I'll add the missing statuses and schedule a release later this week.

**Comment by Jericho** [comment]
Actually, someone reported a bug today which causes the nuget package to have reference incorrect version of dependencies so I'll go ahead and release new version this afternoon.

**Comment by tmacht** [comment]
Unfortunately, it seems that the conversion from string to enum actually is case sensitive.

`
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Linq;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Linq;
using NUnit.Framework;

namespace Tests
{
    [TestFixture]
    public class Tests
    {
        [Test]
        public void TestBla()
        {
            // Works
            //var obj = JObject.Parse("{ \"Status\": \"in progress\" }");

            // Causes exception
            var obj = JObject.Parse("{ \"Status\": \"In progress\" }");
            var res = obj["Status"].ToObject<CampaignStatus>();
            int xx = 9;
        }

        /// <summary>
        /// Enumeration to indicate the status of a campaign
        /// </summary>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum CampaignStatus
        {
            /// <summary>
            /// The campaign has been canceled
            /// </summary>
            [EnumMember(Value = "canceled")]
            Canceled,

            /// <summary>
            /// The campaign is in draft mode
            /// </summary>
            [EnumMember(Value = "draft")]
            Draft,

            /// <summary>
            /// The campaign is beeing delivered
            /// </summary>
            [EnumMember(Value = "in progress")]
            InProgress,

            /// <summary>
            /// The campaign has been scheduled
            /// </summary>
            [EnumMember(Value = "scheduled")]
            Scheduled,

            /// <summary>
            /// The campaign has been sent
            /// </summary>
            [EnumMember(Value = "sent")]
            Sent
        }
    }
}
`

**Comment by Jericho** [comment]
I don't remember what tests I ran back in October that led me to conclude that StringEnumConverter was case-insensitive but, as you demonstrated, that's not the case. I improved my unit test to test multiple casing and only the first test succeeds which further proves that the conversion is case sensitive. 

```csharp
[Theory]
[InlineData("in progress", CampaignStatus.InProgress)]
[InlineData("IN PROGRESS", CampaignStatus.InProgress)]
[InlineData("In Progress", CampaignStatus.InProgress)]
[InlineData("In progress", CampaignStatus.InProgress)]
public void Case_insensitive(string json, CampaignStatus expectedStatus)
{
    // Arrange
    var textReader = new StringReader($"'{json}'");
    var jsonReader = new JsonTextReader(textReader);
    var objectType = typeof(CampaignStatus);
    var converter = new StringEnumConverter();

    // Act
    jsonReader.Read();
    var result = converter.ReadJson(jsonReader, objectType, (object)null, new JsonSerializer());

    // Assert
    result.ShouldNotBeNull();
    result.ShouldBeOfType<CampaignStatus>();
    ((CampaignStatus)result).ShouldBe(expectedStatus);
}
```

I'll research how I can ensure the conversion is case-insensitive.

**Comment by Jericho** [comment]
Ok, turns out Json.net StringEnumConverter **used** to be case insensitive but that was changed in version 10. Here's the discussion that explains [why the case-sensitivity was changed](https://github.com/JamesNK/Newtonsoft.Json/issues/1067) and here's the [commit](https://github.com/JamesNK/Newtonsoft.Json/commit/b2afa9d674f4dc0d5137ce6f175e06e565fc66f5) that actually changed the behavior. StrongGrid used Json.net v9 until v0.28.0 and the Json.net was upgrade to v10 in StrongGrid v0.29.0. That probably explains why I thought the conversion was case insensitive.

**Comment by Jericho** [comment]
I started a [discussion](https://github.com/JamesNK/Newtonsoft.Json/issues/1516) in the Json.net github project and also found [another discussion](https://github.com/JamesNK/Newtonsoft.Json/issues/1426) regarding this issue.

**Comment by Jericho** [comment]
Sorry it's taking so long. Just want to keep you in the loop:

To address once and for all the issue you have raised, I have submitted a [PR to the Json.net project](https://github.com/JamesNK/Newtonsoft.Json/pull/1519) which would restore case-insensitivity when converting a string to `Enum`. Unfortunately, @jamesNK (the author of the Json.net project) decided to decline my submission but the great news is that he decided to revisit the logic of the `StringEnumConverter` and address the case-sensitivity situation.

I will update StrongGrid as soon as the next version of Json.net becomes available. In the mean time, I have added unit tests to StrongGrid so we can validate case-insensitivity.
